### PR TITLE
Workspace leaks

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs [hunk 1]
```diff
@@ -43,17 +43,11 @@ namespace MonoDevelop.CSharp.OptionProvider
...
-
-		public CSharpDocumentOptionsProvider (Workspace workspace)
-		{
-			this.workspace = workspace;
-		}
-
 		async Task<IDocumentOptions> IDocumentOptionsProvider.GetOptionsForDocumentAsync (Document document, CancellationToken cancellationToken)
 		{
-			var mdws = (MonoDevelopWorkspace)workspace;
+			var mdws = document?.Project?.Solution?.Workspace as MonoDevelopWorkspace;
```
Comment (mkrueger) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs]: A rosyln document has always a project - so it's a bit too much.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs [hunk 2]
```diff
@@ -81,7 +75,7 @@ static string GetPath(Document document)
 				return document.FilePath;
 
 			// The file might not actually have a path yet, if it's a file being proposed by a code action. We'll guess a file path to use
-			if (document.Name != null && document.Project.FilePath != null) {
+			if (document.Name != null && document?.Project.FilePath != null) {
```
Comment (mkrueger) [inline main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs]: if document is null that's not working.

File: main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs (hunk 1 above)
Comment (Therzok) [reply main/src/addins/CSharpBinding/MonoDevelop.CSharp.OptionProvider/CSharpDocumentOptionsProvider.cs]: Updated.

**Comment by Therzok** [comment]
Rebased, will need another review.

**Comment by Therzok** [comment]
```
2018-04-10T18:32:21.9341620Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/ConsoleProject.sln-272/ConsoleProject/ConsoleProject.csproj': Project does not support framework '.NETCoreApp,Version=v1.0'
2018-04-10T18:33:09.4687860Z WARNING: Could not load project 'PortableTest.csproj' with unknown item type '{786C830F-07A1-408B-BD7F-6EE04809D6DB}'
2018-04-10T18:33:09.4716800Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/PortableTest.sln-317/PortableTest/PortableTest.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
2018-04-10T18:33:09.4811060Z WARNING: Could not load project 'PortableTest.csproj' with unknown item type '{786C830F-07A1-408B-BD7F-6EE04809D6DB}'
2018-04-10T18:33:09.4839670Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/PortableTest.sln-318/PortableTest/PortableTest.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
2018-04-10T18:33:09.4944630Z WARNING: Could not load project 'PortableTest.csproj' with unknown item type '{786C830F-07A1-408B-BD7F-6EE04809D6DB}'
2018-04-10T18:33:09.4981820Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/PortableTest.sln-319/PortableTest/PortableTest.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
2018-04-10T18:33:09.5230320Z WARNING: Could not load project 'PortableTest.csproj' with unknown item type '{786C830F-07A1-408B-BD7F-6EE04809D6DB}'
2018-04-10T18:33:09.5259050Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/PortableTest.sln-320/PortableTest/PortableTest.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
2018-04-10T18:33:25.6366610Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/DotNetCoreSharedProjectTest.sln-364/DotNetCore/DotNetCoreProject.csproj': Project does not support framework '.NETCoreApp,Version=v1.0'
2018-04-10T18:33:43.8213490Z WARNING: Could not load Workflow Foundation project 'library1.csproj'. This project type is not supported by Visual Studio 2017 for Mac (Preview).
2018-04-10T18:33:43.9353780Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/console-with-libs.sln-378/library1/library1.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
```

Is the bot badly provisioned?

**Comment by Therzok** [comment]
Also:
```
2018-04-10T18:27:23.3137170Z    Loading projects ..
2018-04-10T18:27:33.6673820Z WARNING: Error trying to read configure.in ('/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop') for project 'ConsoleProject':
2018-04-10T18:27:33.6699660Z Unable to find configure.in at '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/configure.ac'.
```
```
2018-04-10T18:29:11.3556500Z WARNING: Could not load Workflow Foundation project 'library1.csproj'. This project type is not supported by Visual Studio 2017 for Mac (Preview).
2018-04-10T18:29:11.3687410Z WARNING: Error while trying to load the project '/Users/vsts/agent/2.131.0/work/r1/a/VS for Mac/drop/tests/tmp/console-with-libs.sln-166/library1/library1.csproj': Unknown solution item type: {FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}
```

**Comment by Therzok** [comment]
Seems like it's an intermittent failure: https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=1574798

**Comment by slluis** [comment]
@mkrueger can you approve?

**Pull request merged**